### PR TITLE
qt: add Vulkan/molten-vk support

### PR DESCRIPTION
**Pull request by loganmc10**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
Hi, I'm hoping to resurrect https://github.com/Homebrew/homebrew-core/pull/50038, I'm not sure why it wasn't merged. See also https://www.qt.io/blog/2018/05/30/vulkan-for-qt-on-macos

I am the developer of simple64 (https://simple64.github.io), an N64 emulator. I'd like to add Mac OS support, and this is blocking me

### CODE REVIEW

**Comment by loganmc10** [comment]
@carlocab I see that you've been handling recent Qt commits

File: Formula/qt.rb [hunk 1]
```diff
@@ -47,6 +47,7 @@ class Qt < Formula
   depends_on "libpng"
   depends_on "libtiff"
   depends_on "md4c"
+  depends_on "molten-vk"
```
Comment (SMillerDev) [inline Formula/qt.rb:50]: Is this a runtime or a build time dependency?

File: Formula/qt.rb [hunk 2]
```diff
@@ -166,6 +167,9 @@ def install
       -no-sql-mysql
       -no-sql-odbc
       -no-sql-psql
+
+      -I#{Formula["molten-vk"].opt_include}
+      -I#{Formula["molten-vk"].opt_libexec}/include
```
Comment (SMillerDev) [inline Formula/qt.rb:172]: It needs to detect dependencies on it's own, this is not a proper solution

**Comment by SMillerDev** [comment]
Please don't tag individual maintainers, no pull request is not important enough to demand volunteer time like that.

**Comment by carlocab** [comment]
I don't have the time to review this properly, but a few comments:

1. Commit message needs fixing. (See [guide](https://docs.brew.sh/Formula-Cookbook#commit) in PR template or summary below.)
2. This dependency needs to be macOS-only, since MoltenVK is not available on Linux.
3. MoltenVK is a pretty heavy dependency to add for something that appears pretty niche.

RE commit style:

Please use the preferred commit-message style for homebrew/core. We put the name of the formula first in commit-message headings.

For new formulae:

> At Homebrew, we like to put the name of the formula up front like so: `foobar 7.3 (new formula)`.

For existing formulae:

> The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

Refer to the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit) for more details. Also, when making further changes to your pull request, use the following guidelines to make sure that @BrewTestBot can merge your commits:

> - One formula per commit; one commit per formula.
> - Keep merge commits out of the pull request.

File: Formula/qt.rb (hunk 1 above)
Comment (loganmc10) [reply Formula/qt.rb:50]: Both

File: Formula/qt.rb (hunk 2 above)
Comment (loganmc10) [reply Formula/qt.rb:172]: It's the only way it works, and it the recommended way according to this: https://www.qt.io/blog/2018/05/30/vulkan-for-qt-on-macos

**Comment by loganmc10** [comment]
> Commit message needs fixing. (See [guide](https://docs.brew.sh/Formula-Cookbook#commit) in PR template or summary below.)

Done
> This dependency needs to be macOS-only, since MoltenVK is not available on Linux.

Done
> MoltenVK is a pretty heavy dependency to add for something that appears pretty niche.

Qt is primarily used for GUI applications, and with OpenGL deprecated on Mac OS, it's really the only good option for GPU accelerated applications. If you install Qt on a Linux distro, or Windows, it's going to come with Vulkan support, so excluding it on brew causes a significant difference in features vs Linux or Windows.

File: Formula/qt.rb (hunk 2 above)
Comment (Bo98) [reply Formula/qt.rb]: Should work without if `find_package(Vulkan)` works. I'm guessing that doesn't?

**Comment by carlocab** [comment]
Thoughts here, @paperchalice, @jonaski?

**Comment by carlocab** [comment]
> excluding it on brew causes a significant difference in features vs Linux or Windows.

This doesn't really bother me -- having fewer features than on Linux is pretty typical for macOS. To me, it's really a matter of whether there will be a significant fraction of users of this formula who would find this useful.

**Comment by paperchalice** [comment]
`qt` with vulkan support in fact doesn't contain the `molten-vk` linkage, so mark it as build dependency and introduce build dependency `vulkan-headers` is enough, and the extra cmake flags are unnecessary here. For test, we need `vulkan-loader` and a vulkan implementation. BTW there is an [upstream issue](https://bugreports.qt.io/browse/QTBUG-106912) about it on macOS.

**Comment by carlocab** [comment]
> `qt` with vulkan support in fact doesn't contain the `molten-vk` linkage, so mark it as build dependency and introduce build dependency `vulkan-headers` is enough, and the extra cmake flags are unnecessary here. For test, we need `vulkan-loader`, `vulkan-headers` and a vulkan implementation. BTW there is an [upstream issue](https://bugreports.qt.io/browse/QTBUG-106912) about it on macOS.

If it's a build-only dependency then I'm fine with this. We do need a test, though; thanks for laying out a starting point.

**Comment by loganmc10** [comment]
Full disclosure: I don't have a Mac or any way to test this, so I'm fine with whatever solution as long as Vulkan support is added to Qt, I'd like to be able to provide a Mac build of the emulator

File: Formula/qt.rb [hunk 3]
```diff
@@ -186,6 +190,8 @@ def install
       # NOTE: `chromium` should be built with the latest SDK because it uses
       # `___builtin_available` to ensure compatibility.
       config_args << "-skip" << "qtwebengine" if DevelopmentTools.clang_build_version <= 1200
+      config_args << "-I#{Formula["molten-vk"].opt_include}"
+      config_args << "-I#{Formula["molten-vk"].opt_libexec}/include"
```
Comment (paperchalice) [inline Formula/qt.rb:194]:
```suggestion
```

File: Formula/qt.rb [hunk 4]
```diff
@@ -64,6 +64,10 @@ class Qt < Formula
   uses_from_macos "libxslt"
   uses_from_macos "zlib"
 
+  on_macos do
+    depends_on "molten-vk"
```
Comment (paperchalice) [inline Formula/qt.rb:68]:
```suggestion
    depends_on "molten-vk" => [:build, :test]
```
Then add generic dependencies:
```ruby

    depends_on "vulkan-headers" => [:build, :test]
    depends_on "vulkan-loader" => :test
```
On linux we can use vulkan icd from `mesa`.

**Comment by loganmc10** [comment]
ok PR has been updated per @paperchalice suggestions

**Comment by carlocab** [comment]
> Full disclosure: I don't have a Mac or any way to test this, so I'm fine with whatever solution as long as Vulkan support is added to Qt, I'd like to be able to provide a Mac build of the emulator

Does this mean you haven't been able to test whether changing our Qt build in this way works for your changes at simple64/simple64#316?

If we add a test for Vulkan support I can cherry-pick this onto #110816 and get a build going.

**Comment by loganmc10** [comment]
> Does this mean you haven't been able to test whether changing our Qt build in this way works for your changes at [simple64/simple64#316](https://github.com/simple64/simple64/pull/316)?

Correct, simple64 used to support Mac, another developer worked on the support, but it stopped working when we migrated from OpenGL to Vulkan. I'm trying to get a DMG again so that someone can test it and get it working if anything needs to be done.

**Comment by carlocab** [comment]
Ok, given that we don't even know if this change works for anybody, I'm going to insist that we have a test for this functionality. You may need to find a Mac to build and test on -- GitHub runners will not suffice (the time to build Qt exceeds their 6 hour time limit).

Or, perhaps @paperchalice might be interested in picking this up.

**Comment by carlocab** [comment]
I don't think there's going to be any progress here, so I'm closing this.

If you'd still like to pursue this change, please feel free to open a new pull request. However, if you choose to do so: please test that your changes produce the intended result, and update the formula's test to include a minimal check for the usability of the new features.

**Pull request closed without merging**